AIP-61: Allow multiple executors to load their CLI commands

syedahsn opened the pull request:
Executors are able to vend custom CLI commands to the Airflow CLI. In order to support multiple Executors, we need to allow each configured Executor to vend its set of CLI commands. 

This PR introduces a method to load all Executor specific CLI commands, as well as associated unit tests. 

<!--
 Licensed to the Apache Software Foundation (ASF) under one
 or more contributor license agreements.  See the NOTICE file
 distributed with this work for additional information
 regarding copyright ownership.  The ASF licenses this file
 to you under the Apache License, Version 2.0 (the
 "License"); you may not use this file except in compliance
 with the License.  You may obtain a copy of the License at

   http://www.apache.org/licenses/LICENSE-2.0

 Unless required by applicable law or agreed to in writing,
 software distributed under the License is distributed on an
 "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 KIND, either express or implied.  See the License for the
 specific language governing permissions and limitations
 under the License.
 -->

<!--
Thank you for contributing! Please make sure that your code changes
are covered with tests. And in case of new features or big changes
remember to adjust the documentation.

Feel free to ping committers for the review!

In case of an existing issue, reference it using one of the following:

closes: #ISSUE
related: #ISSUE

How to write a good git commit message:
http://chris.beams.io/posts/git-commit/
-->



<!-- Please keep an empty line above the dashes. -->
---
**^ Add meaningful description above**
Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/contributing-docs/05_pull_requests.rst#pull-request-guidelines)** for more information.
In case of fundamental code changes, an Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvement+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in a newsfragment file, named `{pr_number}.significant.rst` or `{issue_number}.significant.rst`, in [newsfragments](https://github.com/apache/airflow/tree/main/newsfragments).

syedahsn committed:
Allow multiple executors to load their CLI commands
Unit tests

syedahsn commented on airflow/cli/cli_parser.py:
```diff
@@ -90,8 +93,7 @@
         f"The following CLI {len(dup)} command(s) are defined more than once: {sorted(dup)}\n"
-        f"This can be due to the executor '{ExecutorLoader.get_default_executor_name()}' "
-        f"redefining core airflow CLI commands."
+        f"This can be due to an Executor or Auth Manager redefining core airflow CLI commands."
```
Auth managers can also [vend](https://airflow.apache.org/docs/apache-airflow/stable/core-concepts/auth-manager.html#cli) CLI commands, so this message needs to be updated as well.

o-nikolas commented on airflow/cli/cli_parser.py:
```diff
@@ -56,19 +56,22 @@
+
+executors = [executor for executor, _ in ExecutorLoader.import_all_executors()]
+
+for executor in executors:
```
You want the imports to be within the try/catch here, which is quite important. The biggest issue we're looking out for here is the case of the executor module not being able to load (maybe it has some dependency that's not installed) which is what you'd see if you configured an executor in airflow config but didn't install the provider package say.

With this in mind, you might have to pivot to adding a method to just fetch the names, and then load them individually so that one failed executor doesn't stop others from loading their commands.

So the pseudo code will look like:
```python
for executor_name in ExecutorLoader.get_all_executor_names():
   try:
       # Both the below lines need to be in the try
       loaded_executor = ExecutorLoader.load_executor(executor_name)
       airflow_commands.extend(loaded_executor.get_cli_commands())
   except Exception:
       <all the same exception handling/messaging here>

syedahsn committed:
Load executors inside of a Try/except block to prevent failures if an Executor fails to load
Add unit test to test failure to load Executor case
Remove import_all_executors and associated unit test
